design-documents/graph-ql/coverage/returns.graphqls

```diff
@@ -0,0 +1,197 @@
+type Query{
+    # See https://github.com/magento/architecture/blob/master/design-documents/graph-ql/framework/attributes-metadata.md
+    pageSpecificCustomAttributes(
+        page_type: CustomAttributesPageEnum!
+    ): CustomAttributeMetadata
+}
+
+type Mutation {
+    createReturn(input: CreateReturnInput!): CreateReturnOutput @doc(description: "Create a new return.")
+    addReturnComment(input: AddReturnCommentInput!): AddReturnCommentOutput @doc(description: "Add a comment to an existing return.")
+    addReturnTracking(input: AddReturnTrackingInput!): AddReturnTrackingOutput @doc(description: "Add tracking information to the return.")
+    removeReturnTracking(input: RemoveReturnTrackingInput!): RemoveReturnTrackingOutput @doc(description: "Remove tracking information from the return.")
+}
+
+input CreateReturnInput {
+    items: [CreateReturnItemInput]!
+    comment_text: String
+}
+
+input CreateReturnItemInput {
+    order_item_id: ID!
+    quantity_to_return: Float!
+    selected_custom_attributes: [ID!]
+    entered_custom_attributes: [EnteredCustomAttributeInput!]
+}
+
+input EnteredCustomAttributeInput {
+    id: ID!
+    value: String!
+}
+
+type CreateReturnOutput {
+    return: Return
+}
+
+input AddReturnCommentInput {
+    return_id: ID!
+    comment_text: String!
+}
+
+type AddReturnCommentOutput {
+    return: Return
+}
+
+input AddReturnTrackingInput {
+    return_id: ID!
+    carrier_id: ID!
+    tracking_number: String!
+}
+
+type AddReturnTrackingOutput {
+    return: Return
```

Comment:
It would be nice if this also included `tracking: ReturnShippingTracking` - client libraries will almost always want to be able to get to the `ID` (at least) of a resource that was just created.

Reply:
See example below (took it from .md file). Isn't this enough?

```graphql
mutation {
  addReturnTracking(
    input: {
      return_id: "000005", 
      carrier_id: "carrier-id", 
      tracking_number: "4234213"
    }
  ) {
    return {
      shipping {
        tracking {
          id
          carier {
            label
          }
          tracking_number
        }
      }
    }
  }
}
```

Reply:
Isn't `data.return.shipping.tracking` a list? The client would have to filter this list on `tracking_number`, and then update the cache based on that.

Ideally, I think all mutations would at minimum give easy access to the resource it is intended to create.

Reply:
I would like to the full `Return` object for maximum flexibility. 

We can add `ReturnShippingTracking` to the mutation output, but it seems excessive. 
Considering that multiple tracking numbers will be rare I don't see a problem in refreshing the whole list of tracking numbers after the mutation.

Reply:
> I would like to have the full Return object for maximum flexibility.

Not suggesting we get rid of the `Return` object, just suggesting we extend the output type.

If a client is adding an item to a collection, I'd expect it to be fairly straight-forward for the client to access just that resource if it wants, without having to grab extra data and filter.

-----

https://www.apollographql.com/docs/react/data/mutations/#making-all-other-cache-updates

> If a mutation modifies multiple entities, or **if it creates or deletes entities**, the Apollo Client cache is not automatically updated to reflect the result of the mutation. To resolve this, your call to useMutation can include an update function.

Reply:
Added